sdk/examples/xo_java/XoHandler.java

```diff
@@ -0,0 +1,445 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Logger;
+
+public class XoHandler implements TransactionHandler {
+
+  private final Logger logger = Logger.getLogger(XoHandler.class.getName());
+  private String xoNameSpace;
+
+  /**
+   * constructor.
+   */
+  public XoHandler() {
+    try {
+      this.xoNameSpace = Utils.hash512(
+        this.transactionFamilyName().getBytes("UTF-8")).substring(0, 6);
+    } catch (UnsupportedEncodingException usee) {
+      usee.printStackTrace();
+      this.xoNameSpace = "";
+    }
+  }
+
+  @Override
+  public String transactionFamilyName() {
+    return "xo";
+  }
+
+  @Override
+  public String getEncoding() {
+    return "csv-utf8";
+  }
+
+  @Override
+  public String getVersion() {
+    return "1.0";
+  }
+
+  @Override
+  public Collection<String> getNameSpaces() {
+    ArrayList<String> namespaces = new ArrayList<>();
+    namespaces.add(this.xoNameSpace);
+    return namespaces;
+  }
+
+  @Override
+  public void apply(TpProcessRequest transactionRequest, State stateStore)
+      throws InvalidTransactionException, InternalError {
+    String[] transactionData = getUnpackedTransaction(transactionRequest);
+    String action = transactionData[0];
+    String gameName = transactionData[1];
+    String space = transactionData[2];
+
+    if (gameName.equals("")) {
+      throw new InvalidTransactionException("Name is required");
+    }
+
+    if (action.equals("")) {
+      throw new InvalidTransactionException("Action is required");
+    }
+
+    String[] stateData = getStateData(gameName, stateStore);
+    String board = stateData[0];
+    String state = stateData[1];
+    String player1 = stateData[2];
+    String player2 = stateData[3];
+
+    String signer = getSignerPubkey(transactionRequest);
+    String[] updatedGameData = playXo(
+      gameName, board, state,
+      player1, player2,
+      signer, action, space
+    );
+
+    storeGameData(gameName, updatedGameData, stateStore);
+  }
+
+  /**
+   * Helper function to get signer public key from transaction header.
+   */
+  private String getSignerPubkey(TpProcessRequest transactionRequest) throws InternalError {
+    try {
+      TransactionHeader header = TransactionHeader.parseFrom(transactionRequest.getHeader());
+      return header.getSignerPubkey();
+    } catch (InvalidProtocolBufferException e) {
+      throw new InternalError("Protocol Buffer Error: " + e.toString());
+    }
+  }
+
+  /**
+   * Helper function to retrieve game name, action, and space from transaction request.
+   */
+  private String[] getUnpackedTransaction(TpProcessRequest transactionRequest) {
+    String[] transactionData = {"", "", ""};
+    String[] payload =  csvDecodeData(transactionRequest.getPayload().toStringUtf8());
+    System.arraycopy(payload, 0, transactionData, 0, payload.length);
```

Comment:
This `arraycopy` is unnecessary, as the underlying split is creating a new array every time. Likewise, if you create the data object class, as mentioned above, you can return an new instance of one of those here.